### PR TITLE
Use the REST API for Telemetry

### PR DESCRIPTION
**Pull request by Snuffy2**
### Missing

* OpenVPN _(I don't use OpenVPN so I don't have a good way to test this)_
    * Connected Client Count
* CPU
    * Frequency Current
    * Frequency Max

### OPNSense User Permissions Needed

* Dashboard (all)
* Status: Interfaces
* Lobby: Login / Logout / Dashboard
* Status: OpenVPN
* VPN: OpenVPN: Client Export Utility 

### API Endpoints Used

| Endpoint | Permission |
| -------- | ------- |
| /api/interfaces/overview/export | Status: Interfaces |
| /api/diagnostics/system/system_mbuf | Dashboard (all) |
| /api/diagnostics/firewall/pfstates | Dashboard (all) |
| /api/diagnostics/system/systemResources | Dashboard (all) |
| /api/diagnostics/system/systemSwap | Dashboard (all) |
| /api/diagnostics/system/systemTime | Lobby: Login / Logout / Dashboard |
| /api/diagnostics/cpu_usage/getCPUType | Dashboard (all) |
| /api/diagnostics/system/systemDisk | Dashboard (all) |
| /api/openvpn/export/providers | VPN: OpenVPN: Client Export Utility |
| /api/openvpn/service/searchSessions | Status: OpenVPN |
| /api/routes/gateway/status | Dashboard (all) |

May be a replacement/fix for #143

### CODE REVIEW

**Comment by alexdelprete** [comment]
@Snuffy2 I wanted to test it. But I have the following warnings/errors on HA restart:

image

image

And this one is preventing the integration to initialize:

```console
Logger: custom_components.opnsense
Source: helpers/update_coordinator.py:354
integration: OPNsense ([documentation](https://github.com/travisghansen/hass-opnsense), [issues](https://github.com/travisghansen/hass-opnsense/issues))
First occurred: 04:59:03 (6 occurrences)
Last logged: 05:01:15

Unexpected error fetching OPNsense OPNsense state data
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 271, in _async_update_data
    return await self.update_method()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/__init__.py", line 97, in async_update_data
    await hass.async_add_executor_job(lambda: data.update())
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/__init__.py", line 97, in <lambda>
    await hass.async_add_executor_job(lambda: data.update())
                                              ^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/__init__.py", line 316, in update
    self._state["telemetry"] = self._get_telemetry()
                               ^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/__init__.py", line 225, in inner
    response = func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/__init__.py", line 249, in _get_telemetry
    return self._client.get_telemetry()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 103, in inner
    raise err
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 100, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 867, in get_telemetry
    telemetry["openvpn"] = self._get_telemetry_openvpn()
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 103, in inner
    raise err
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 100, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/opnsense/pyopnsense/__init__.py", line 1016, in _get_telemetry_openvpn
    for vpnid, vpn_info in openvpn_info.items():
                           ^^^^^^^^^^^^^^^^^^
AttributeError: 'list' object has no attribute 'items'
```

**Comment by alexdelprete** [comment]
This is from the debug log:

```log
2024-09-02 05:01:27.035 DEBUG (SyncWorker_18) [custom_components.opnsense.pyopnsense] [get_telemetry_openvpn] openvpn_info: []
2024-09-02 05:01:27.035 DEBUG (SyncWorker_18) [custom_components.opnsense.pyopnsense] [get_telemetry_openvpn] connection_info: {'total': 0, 'rowCount': 0, 'current': 1, 'rows': []}
2024-09-02 05:01:27.035 ERROR (SyncWorker_18) [custom_components.opnsense.pyopnsense] Unexpected _get_telemetry_openvpn error err=AttributeError("'list' object has no attribute 'items'"), type(err)=<class 'AttributeError'>
2024-09-02 05:01:27.035 ERROR (SyncWorker_18) [custom_components.opnsense.pyopnsense] Unexpected get_telemetry error err=AttributeError("'list' object has no attribute 'items'"), type(err)=<class 'AttributeError'>
2024-09-02 05:01:27.035 ERROR (MainThread) [custom_components.opnsense] Unexpected error fetching OPNsense OPNsense state data
```

File: custom_components/opnsense/pyopnsense/__init__.py [hunk 1]
```diff
@@ -853,193 +855,195 @@ def send_wol(self, interface, mac):
...
-            data["gateways"] = {}
+    @_log_errors
+    def _get_telemetry_openvpn(self) -> dict:
+        openvpn: dict[str, Any] = {}
+        openvpn["servers"] = {}
+        openvpn_info: dict[str, Any] = self._post("/api/openvpn/export/providers")
+        connection_info: dict[str, Any] = self._post("/api/openvpn/service/searchSessions")
+        _LOGGER.debug(f"[get_telemetry_openvpn] openvpn_info: {openvpn_info}")
+        _LOGGER.debug(f"[get_telemetry_openvpn] connection_info: {connection_info}")
+        for vpnid, vpn_info in openvpn_info.items():
```
Comment (alexdelprete) [inline custom_components/opnsense/pyopnsense/__init__.py:1016]: Added a check if `openvpn_info` list is empty. Happens when there's no OpenVPN interface (at least here), and it stops the integration from loading. I modified it like this and it works now.

```python
    @_log_errors
    def _get_telemetry_openvpn(self) -> dict:
        openvpn: dict[str, Any] = {}
        openvpn["servers"] = {}
        openvpn_info: dict[str, Any] = self._post("/api/openvpn/export/providers")
        connection_info: dict[str, Any] = self._post(
            "/api/openvpn/service/searchSessions"
        )
        _LOGGER.debug(f"[get_telemetry_openvpn] openvpn_info: {openvpn_info}")
        _LOGGER.debug(f"[get_telemetry_openvpn] connection_info: {connection_info}")
        if openvpn_info:
            for vpnid, vpn_info in openvpn_info.items():
                vpn: dict[str, Any] = {}
                vpn["vpnid"] = vpn_info["vpnid"]
                vpn["name"] = vpn_info["name"]
                total_bytes_recv = 0
                total_bytes_sent = 0
                for connect in connection_info["rows"]:
                    if connect["id"] == vpn["vpnid"]:
                        total_bytes_recv += (
                            int(connect["bytes_received"])
                            if connect["bytes_received"]
                            else 0
                        )
                        total_bytes_sent += (
                            int(connect["bytes_sent"]) if connect["bytes_sent"] else 0
                        )
                vpn["total_bytes_recv"] = total_bytes_recv
                vpn["total_bytes_sent"] = total_bytes_sent
                # Missing connected_client_count
                # vpn["connected_client_count"] =
                openvpn["servers"][vpnid] = vpn
            _LOGGER.debug(f"[get_telemetry_openvpn] openvpn: {openvpn}")
        else:
            _LOGGER.debug("[get_telemetry_openvpn] openvpn_info: no OpenVPN interfaces")
        return openvpn
```

File: custom_components/opnsense/pyopnsense/__init__.py [hunk 2]
```diff
@@ -853,193 +855,195 @@ def send_wol(self, interface, mac):
...
+        filesystems: dict[str, Any] = {}
+        filesystems_info: dict[str, Any] = self._post(
+            "/api/diagnostics/system/systemDisk"
+        )
+        _LOGGER.debug(f"[get_telemetry_filesystems] filesystems_info: {filesystems_info}")
+        filesystems = filesystems_info["devices"]
+        # To conform to the previous data being returned
+        for filesystem in filesystems:
+            filesystem["size"] = filesystem.pop("blocks", None)
+            filesystem["capacity"] = f"{filesystem.pop("used_pct")}%"
```
Comment (alexdelprete) [inline custom_components/opnsense/pyopnsense/__init__.py:1004]: I had an issue with VSCode giving me weird errors for this line, had to modify with single quotes:

```python
filesystem["capacity"] = f"{filesystem.pop('used_pct')}%"
```

**Review by alexdelprete — CHANGES_REQUESTED** [review]
A couple of fixes, a minor one and the other was preventing the integration to start.

**Comment by Snuffy2** [comment]
Sorry, lazy coding on my part. Added more robust error handling that hopefully handles most/all potential issues with the data.

**Comment by alexdelprete** [comment]
> I'm seeing this Warning too. Not sure it has anything to do with the code changes in this PR though.
> 
> 

Well, the log is pretty clear, it also mentions the lines: it's always the code `script = """`

image

**Comment by alexdelprete** [comment]
> Sorry, lazy coding on my part. Added more robust error handling that hopefully handles most/all potential issues with the data.

Thank you very much for your contribution. I'll try to understand better how it works so I might be of help in the future. Hope we'll be able to use the REST API for everything in the future. :)

**Comment by Snuffy2** [comment]
I added a commit to handle the Warnings with the PHP scripts. I don't believe they are appearing from any of the code changes in this PR. Instead, I suspect they're now showing due to Python updates in HA 2024.8 . However, making the scripts raw strings should fix it.

**Comment by alexdelprete** [comment]
> I don't believe they are appearing from any of the code changes in this PR

I agree, it's not due to this PR. Sorry if I wrongly attributed those to it and thanks for fixing it, I just tested it and the warning has gone. :)

**Comment by alexdelprete** [comment]
@Snuffy2 the only problem I noticed with this PR is that the interface entities names have changed, instead of wan/lan/optX/etc. now there's the custom name of the interface. It makes sense, but this will require users to do some cleanup work in HA, removing orphan entities with the old names.

**Comment by Snuffy2** [comment]
Yes, it is using the Name instead of the device now. I could change that back if desired but I thought it was more useful with the friendly name. 

Similarly the OpenVPN Server names are now also different because I'm pulling from the new OpenVPN interface and not the legacy one. I can't change that one.

**Comment by alexdelprete** [comment]
> Yes, it is using the Name instead of the device now. I could change that back if desired but I thought it was more useful with the friendly name.

I would use Name as Friendly Name in HA, and Device for the name of the entity. But that's just my humble opinion, let's see what @travisghansen says about this.

> Similarly the OpenVPN Server names are now also different because I'm pulling from the new OpenVPN interface and not the legacy one. I can't change that one.

Obviously we can't do much about that, users will have to adapt.

**Comment by travisghansen** [comment]
Thanks everyone! I am happy to merge this or even better get some additional people added as maintainers on the repo given my limited time.

**Comment by alexdelprete** [comment]
> I am happy to merge this or even better get some additional people added as maintainers on the repo given my limited time.

I'm not an expert developer Travis, but I'd be happy to help as I can. Would be awesome if @Snuffy2 could join the team of maintainers. :)

**Comment by alexdelprete** [comment]
> Yes, it is using the Name instead of the device now. I could change that back if desired but I thought it was more useful with t

I had a chat with Travis, and he agreed with me that it's better changing it back. Do you want to prepare a new PR? Otherwise I'll look at the code and prepare it later. Let me know...

**Comment by Snuffy2** [comment]
@travisghansen, if you don't mind me moving things that I fix or update from XMLRPC/PHP to the REST API using pure Python then I'm happy to help to maintain. Feel free to contact me at github@mail.snuffy2.com if you'd like to contact me outside of GitHub.

As for the interface naming, since v0.1.22 is released and anyone who upgrades to it will get new HA entities, wont it be more confusing to release another update that changes the interface name and IDs again?

**Comment by travisghansen** [comment]
I am happy to let people take the project wherever it needs to go (regarding rest vs xmlrpc).

I think if we’re fast enough it’s fine to catch the id change. Not everyone updates their installs everyday..so a same day release for example to ‘fix’ that issue I think is fine.

**Comment by alexdelprete** [comment]
> I am happy to let people take the project wherever it needs to go (regarding rest vs xmlrpc)

Travis, remember when we asked Franco (OPNsense lead dev) about XMLrpc vs API when we started this adventure? He said eventually XMLrpc will be abandoned in favor of the API, once OPNsense GUI will only use the API.

So I guess the API transition is inevitable. I don't know if the API covers everything that we got via XMLrpc though.

Happy to see that @Snuffy2 accepted to maintain this project. :)

**Comment by alexdelprete** [comment]
Regarding the interface naming: I would use device for the name of entities, and the user description as HA friendly name for the entity. What do you think @Snuffy2?

**Comment by travisghansen** [comment]
xmlrpc was always a temporary solution until the rest api reached parity…when originally built (and likely still) the rest api was limited enough that the whole plugin would be pretty useless, but transitioning as the rest api allows is definitely preferred.

**Comment by alexdelprete** [comment]
and once the API is complete, we can also get rid of the required opnsense plugin.

**Comment by travisghansen** [comment]
Yes that is true.

**Comment by alexdelprete** [comment]
> Feel free to contact me at [github@mail.snuffy2.com](mailto:github@mail.snuffy2.com) if you'd like to contact me outside of GitHub.

Snuffy, in agreement with Travis, we created a Discord server to have a more direct communication channel. I put the link in the readme (the discord badge) if you want to join. :)

Thanks for everything you're doing for this integration.

**Pull request merged**